### PR TITLE
Added join key to OPTIONS, used in dataarray & dataset binary ops, wi…

### PR DESCRIPTION
**Pull request by chunweiyuan**
…th a test module in test_dataarray.py

At the root of the repo, in your new conda environment, run 

> > > pip install -e .

To run unittest on test/test_dataarray.py,

> > > python -m unittest -q test.test_dataarray

For some illustration on how xarray works, please visit
https://stash.ihme.washington.edu/projects/SCIC/repos/pythonutil/browse/tests/test_xarray.py

### CODE REVIEW

File: xarray/core/dataarray.py [hunk 1]
```diff
@@ -1377,13 +1378,15 @@ def func(self, *args, **kwargs):
...
-    def _binary_op(f, reflexive=False, join='inner', **ignored_kwargs):
+    def _binary_op(f, reflexive=False, join=None, **ignored_kwargs):
         @functools.wraps(f)
         def func(self, other):
             if isinstance(other, (Dataset, groupby.GroupBy)):
                 return NotImplemented
             if hasattr(other, 'indexes'):
-                self, other = align(self, other, join=join, copy=False)
+                # if user does not specify join, default to OPTIONS['join']
+                how_to_join = join if join is not None else OPTIONS['join']
```
Comment (vreuter) [inline xarray/core/dataarray.py]: This sort of readability is more of an opinion, but to me this is a little clearer:
`how_to_join = join or OPTIONS["join"]`

File: xarray/core/dataset.py [hunk 2]
```diff
@@ -2018,13 +2019,15 @@ def func(self, *args, **kwargs):
...
     @staticmethod
-    def _binary_op(f, reflexive=False, join='inner', fillna=False):
+    def _binary_op(f, reflexive=False, join=None, fillna=False):
         @functools.wraps(f)
         def func(self, other):
             if isinstance(other, groupby.GroupBy):
                 return NotImplemented
             if hasattr(other, 'indexes'):
-                self, other = align(self, other, join=join, copy=False)
+                # if user does not specify join, default to OPTIONS['join']
```
Comment (vreuter) [inline xarray/core/dataset.py:2028]: could say something like "use default" instead of specifying the mechanism in the comment in case the mechanism changes, but minor

File: xarray/test/test_dataarray.py [hunk 3]
```diff
@@ -2268,3 +2268,38 @@ def test_dot(self):
             da.dot(dm.values)
         with self.assertRaisesRegexp(ValueError, 'no shared dimensions'):
             da.dot(DataArray(1))
+
+    def test_binary_op_join_setting(self):
+        """
+        A test method to verify the ability to set binary operation join kwarg
```
Comment (vreuter) [inline xarray/test/test_dataarray.py:2274]: At least in 2.7, with the way unittest output is displayed, there's a small benefit to having the docstring be single-line since it's easier to see what's under test if there's an error/failure.

File: xarray/core/dataset.py [hunk 4]
```diff
@@ -2018,13 +2019,15 @@ def func(self, *args, **kwargs):
...
+                how_to_join = join if join is not None else OPTIONS['join']
```
Comment (kforeman) [inline xarray/core/dataset.py]: Out of curiosity, why create a new variable `how_to_join`? Convention in [the only other use of `OPTIONS`](https://github.com/pydata/xarray/blob/f40d323c76f6cab9615567e4ba7afde7860f7775/xarray/core/formatting.py#L191-L192) is:

```
if max_width is None:
    max_width = OPTIONS['display_width']
```

**Review by kforeman — COMMENTED** [review]
Looks good! Other than a comment on default `OPTIONS`, nothing on my end.

File: xarray/test/test_dataarray.py [hunk 5]
```diff
@@ -2268,3 +2268,38 @@ def test_dot(self):
...
+        # First we set up a data array
+        xdim, ydim, zdim = 'x', 'y', 'z'
+        xcoords, ycoords, zcoords = ['a', 'b', 'c'], [-2, 0, 2], [0, 1, 2]
+        total_size = len(xcoords) * len(ycoords) * len(zcoords)
+        # create a 3-by-3-by-3 data array
+        arr = xr.DataArray(np.arange(total_size).\
+                           reshape(len(xcoords),len(ycoords),len(zcoords)),
+                           [(xdim, xcoords), (ydim, ycoords),(zdim, zcoords)])
+        # now create a data array with the last x slice missing
+        arr1 = arr[0:-1,:,:].copy()
```
Comment (vreuter) [inline xarray/test/test_dataarray.py]: The comments are good, but you could go with indicative names, too, e.g. `missing_last_x`. I prefer descriptive over terse, but others differ.

File: xarray/test/test_dataarray.py [hunk 6]
```diff
@@ -2268,3 +2268,38 @@ def test_dot(self):
...
+        # create another data array with the last z slice missing
+        arr2 = arr[:,:,0:-1].copy()
+        # because the default in OPTIONS is join="inner", we test "outer" first
+        xr.set_options(join="outer")
+        result = arr1 + arr2
+        self.assertTrue(result.size == total_size) # should be 3 * 3 * 3
+        self.assertTrue(result.shape == arr.shape)
```
Comment (vreuter) [inline xarray/test/test_dataarray.py:2292]: It seems like the size assertion is a given if shape is being asserted, too. Or is there a wonky corner case where you'd need to check both?

File: xarray/test/test_dataarray.py [hunk 7]
```diff
@@ -2268,3 +2268,38 @@ def test_dot(self):
...
+        self.assertTrue(result[-1,:,:].isnull().all())
+        self.assertTrue(result[:,:,-1].isnull().all())
+        # now revert back to join="inner"
+        xr.set_options(join="inner")
```
Comment (vreuter) [inline xarray/test/test_dataarray.py:2296]: Even though it'd mean duplicating some setup code, I would split this larger case into two functions since you're testing two join types. I would say you could make a small class to avoid the duplication, but it looks like pydata has gone with the one-class-per-file approach, at least for these tests. https://docs.python.org/2/library/unittest.html#unittest.TestCase funny given first line

File: xarray/test/test_dataset.py [hunk 8]
```diff
@@ -2916,6 +2917,47 @@ def test_filter_by_attrs(self):
...
+        self.assertTrue(result.foo[:,:,-1].isnull().all())
+        # now revert back to join="inner"
+        xr.set_options(join="inner")
+        result = ds1 + ds2
+        self.assertTrue(result.foo.size == \
+                        (len(xcoords)-1)*len(ycoords)*(len(zcoords)-1))
+        self.assertTrue(result.foo.shape == \
+                        (len(xcoords)-1, len(ycoords), len(zcoords)-1))
+        self.assertTrue(result.foo.notnull().all())
+        self.assertTrue('c' not in list(result.foo['x']))
```
Comment (vreuter) [inline xarray/test/test_dataset.py]: `assertFalse` would be a little clearer here to avoid the negation

**Comment by vreuter** [comment]
Most of my suggestions are opinion/style, overall looks great & it's awesome that you're contributing to open source! Only other thing I'd say is run PEP8 for whitespace since (at least glancing around the tests) it looks like they follow the spacing convention around commas & operators.

File: xarray/core/dataset.py (hunk 4 above)
Comment (chunweiyuan) [reply xarray/core/dataset.py]: Fixed it with a compromise between yours and Vince's comments.  I can't directly copy the max_width case because join is not defined within the wrapped sub-function.

File: xarray/test/test_dataset.py [hunk 9]
```diff
@@ -2932,30 +2932,27 @@ def test_binary_op_join_setting(self):
...
                         (len(xcoords)-1, len(ycoords), len(zcoords)-1))
         self.assertTrue(result.foo.notnull().all())
-        self.assertTrue('c' not in list(result.foo['x']))
-        self.assertTrue(2 not in list(result.foo['z']))
+        self.assertFalse('c' in list(result.foo['x']))
+        self.assertFalse(2 in list(result.foo['z']))
         # just for kicks, what happens when the dataarrays have different names?
-        ds3 = arr1.to_dataset(name='bar')
-        result = ds1 + ds3
+        misnomer = arr1.to_dataset(name='bar')
```
Comment (vreuter) [inline xarray/test/test_dataset.py:2954]: 👍 for name

File: xarray/test/test_dataarray.py [hunk 10]
```diff
@@ -2283,23 +2283,20 @@ def test_binary_op_join_setting(self):
...
-        result = arr1 + arr2
-        self.assertTrue(result.size == \
-                        (len(xcoords)-1)*len(ycoords)*(len(zcoords)-1))
+        result = missing_last_x + missing_last_z
         self.assertTrue(result.shape == \
                         (len(xcoords)-1, len(ycoords), len(zcoords)-1))
         self.assertTrue(result.notnull().all())
-        self.assertTrue('c' not in list(result['x']))
-        self.assertTrue(2 not in list(result['z']))
+        self.assertFalse('c' in list(result['x']))
```
Comment (vreuter) [inline xarray/test/test_dataarray.py:2301]: What's special about '2' and 'c'? (I know these are actually the x and z components that get dropped.) For readability, you could set those between lines 2278 and 2279, though. i.e.:

```
x_drop = 'c'
z_drop = 2
xcoords, ycoords, zcoords = ['a', 'b', x_drop], [-2, 0, 2], [0, 1, z_drop]
```

and then use them to make super explicit what's going on in the test/assertion:

```
self.assertFalse(x_drop in list(result['x']))
etc.
```

**Review by vreuter — APPROVED** [review]
Looks great, just made one more readability suggestion. I'd still recommend running the space checker, but it's your call ;D

**Comment by kheuton** [comment]
Sorry I didn't get to this till tonight, looks good. I'm confused why they chose to not allow outer joins in the first place
